api/v1alpha1/proxydefaults_types.go

```diff
@@ -0,0 +1,196 @@
+}
+
+func (in *ProxyDefaults) SyncedCondition() (status corev1.ConditionStatus, reason, message string) {
+	cond := in.Status.GetCondition(ConditionSynced)
+	return cond.Status, cond.Reason, cond.Message
+}
+
+func (in *ProxyDefaults) SyncedConditionStatus() corev1.ConditionStatus {
+	return in.Status.GetCondition(ConditionSynced).Status
+}
+
+func (in *ProxyDefaults) Name() string {
+	return in.ObjectMeta.Name
+}
+
+func (in *ProxyDefaults) SetSyncedCondition(status corev1.ConditionStatus, reason string, message string) {
+	in.Status.Conditions = Conditions{
+		{
+			Type:               ConditionSynced,
+			Status:             status,
+			LastTransitionTime: metav1.Now(),
+			Reason:             reason,
+			Message:            message,
+		},
+	}
+}
+
+func (in *ProxyDefaults) ToConsul() api.ConfigEntry {
+	consulConfig := in.convertConfig()
+	return &capi.ProxyConfigEntry{
+		Kind:        capi.ProxyDefaults,
+		Name:        in.Name(),
+		MeshGateway: in.Spec.MeshGateway.toConsul(),
+		Expose:      in.Spec.Expose.toConsul(),
+		Config:      consulConfig,
+	}
+}
+
+func (in *ProxyDefaults) MatchesConsul(candidate api.ConfigEntry) bool {
+	proxyDefCand, ok := candidate.(*capi.ProxyConfigEntry)
+	if !ok {
+		return false
+	}
+	return in.Name() == proxyDefCand.Name &&
+		in.Spec.MeshGateway.Mode == string(proxyDefCand.MeshGateway.Mode) &&
+		in.Spec.Expose.matches(proxyDefCand.Expose) &&
+		in.matchesConfig(proxyDefCand.Config)
+}
+
+func (in *ProxyDefaults) Validate() error {
+	var allErrs field.ErrorList
+	path := field.NewPath("spec")
+
+	if err := in.Spec.MeshGateway.validate(path.Child("meshGateway")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	if err := in.validateConfig(path.Child("config")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+	allErrs = append(allErrs, in.Spec.Expose.validate(path.Child("expose"))...)
+	if len(allErrs) > 0 {
+		return apierrors.NewInvalid(
+			schema.GroupKind{Group: ConsulHashicorpGroup, Kind: "proxydefaults"},
+			in.Name(), allErrs)
+	}
+
+	return nil
+}
+
+// matchesConfig compares the values of the config on the spec and that on the
+// the consul proxy-default and returns true if they match and false otherwise
+func (in *ProxyDefaults) matchesConfig(config map[string]interface{}) bool {
+	if in.Spec.Config == nil || config == nil {
+		return in.Spec.Config == nil && config == nil
+	}
+	var inConfig map[string]interface{}
+	if err := json.Unmarshal(in.Spec.Config, &inConfig); err != nil {
+		return false
+	}
+	return reflect.DeepEqual(inConfig, config)
+}
+
+// convertConfig converts the config of type json.RawMessage which is stored
+// by the resource into type map[string]interface{} which is saved by the
+// consul API.
+func (in *ProxyDefaults) convertConfig() map[string]interface{} {
+	if in.Spec.Config == nil {
+		return nil
+	}
+	var outConfig map[string]interface{}
+	if err := json.Unmarshal(in.Spec.Config, &outConfig); err != nil {
+		return nil
+	}
+	return outConfig
```

Comment:
```
// If a JSON value is not appropriate for a given target type,
// or if a JSON number overflows the target type, Unmarshal
// skips that field and completes the unmarshaling as best it can.
// If no more serious errors are encountered, Unmarshal returns
// an UnmarshalTypeError describing the earliest such error. In any
// case, it's not guaranteed that all the remaining fields following
// the problematic one will be unmarshaled into the target object.
```
I could be misreading this but it appears as though, Unmarshall does best effort unmarshalling and we probably dont want to work with an incomplete struct.

Reply:
👍 good find.